key-share/src/valid.rs

```diff
@@ -47,13 +122,19 @@ where
         }
     }
 
-    /// Constructs `Self` from its [`ValidProjection`]
-    pub fn from<K>(value: Valid<K>) -> Self
-    where
-        T: ValidProjection<K>,
-        K: Validate,
-    {
-        Self(value.0.into())
+    /// Constructs `&Valid<T>` from `&T`, assumes that `T` has been validated
+    ///
+    /// Performs a debug assertion that `T` is validated
+    fn from_ref_unchecked(value: &T) -> &Self {
+        #[cfg(debug_assertions)]
+        value
+            .is_valid()
+            .expect("debug assertions: value is invalid, but was assumed to be valid");
+
+        // SAFETY: &T and &Valid<T> have exactly the same in-memory representation
+        // thanks to `repr(transparent)`, so it's sound to transmute the references.
+        // Note also that input and output references have exactly the same lifetime.
+        unsafe { core::mem::transmute(value) }
```

Comment:
Maybe for debug assertions check the resulting value for validity as well?

Reply:
Oh, actually, I meant to check correctness of the output

Reply:
It probably doesn't make much sense to check before conversion

Reply:
Wait, but it does check correctness of output value when doing conversion though `as_ref`

Reply:
This method doesn't coerce the type, it only changes `&T -> &Valid<T>`

Reply:
`sub_value` has type `&K`, so `&K` will be validated for correctness by `from_ref_unchecked`

https://github.com/dfns/cggmp21/blob/2b2ec834ee55816e5feb7d555340d9d8ba7500e1/key-share/src/valid.rs#L162-L171

Reply:
Right, I just got confused as to what Self and T refer to. Everything is good here.